include/oneapi/dpl/internal/binary_search_impl.h

```diff
@@ -134,13 +138,29 @@ lower_bound_impl(__internal::__hetero_tag<_BackendTag>, Policy&& policy, InputIt
     auto keep_result = oneapi::dpl::__ranges::__get_sycl_range<__bknd::access_mode::read_write, OutputIterator>();
     auto result_buf = keep_result(result, result + value_size);
     auto zip_vw = make_zip_view(input_buf.all_view(), value_buf.all_view(), result_buf.all_view());
-    __bknd::__parallel_for(_BackendTag{}, ::std::forward<Policy>(policy),
-                           custom_brick<StrictWeakOrdering, decltype(size), search_algorithm::lower_bound>{comp, size},
-                           value_size, zip_vw)
-        .wait();
+    auto run_lower_bound = [comp, value_size, zip_vw](auto&& policy, auto size) {
+        __bknd::__parallel_for(
+            _BackendTag{}, ::std::forward<decltype(policy)>(policy),
+            custom_brick<StrictWeakOrdering, decltype(size), search_algorithm::lower_bound>{comp, size}, value_size,
+            zip_vw)
+            .wait();
+    };
+    // Enable index calculation to proceed with uint32_t if input range is small enough.
+    if (size <= ::std::numeric_limits<::std::uint32_t>::max())
+    {
+        run_lower_bound(::std::forward<Policy>(policy), static_cast<::std::uint32_t>(size));
+    }
+    else
+    {
+        run_lower_bound(__bknd::make_wrapped_policy<lower_bound_fallback>(::std::forward<Policy>(policy)),
+                        static_cast<::std::size_t>(size));
+    }
```

Comment:
Have you tried to move the runtime decision to `custom_brick` instead? As far as I can see, the size value to make the decision is the same on the host and in the brick, so that should be technically possible. How would that perform?

Reply:
I did some experimenting and there seems to be a small performance penalty to performing the check within the kernel itself (<1% for large n and large number of search keys). However, this allows us to have a compiled single kernel which is likely worth it.

I will do a bit more investigation of doing this runtime check in the kernel.

Reply:
As described at https://en.cppreference.com/w/cpp/types/size_t, "The bit width of std::size_t is not less than 16.(since C++11)"
So does this solution quite correct when `size > ::std::numeric_limits<::std::uint32_t>::max()` ?
Is OK that n this case you simple using `static_cast<::std::size_t>(size)` ?
May be we should use some another type instead of `std::size_t` ?

Reply:
I see your point here. I had used `size_t` for the opposite case where it could holder larger values than `uint64_t`. In other places in the code where we do this, I see we use `std::uint64_t` as the fallback. 

Does switching to `std::uint64_t` seem like a good alternative to you?

Reply:
Probably yes.